include/artifact-uploader.h

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (C) 2017 Prism Skylabs
+ */
+#ifndef CONNECT_SDK_ARTIFACT_UPLOADER_H
+#define CONNECT_SDK_ARTIFACT_UPLOADER_H
+
+#include "domain-types.h"
+#include "boost/shared_ptr.hpp"
+#include "public-util.h"
+
+namespace prism
+{
+namespace connect
+{
+
+class Client;
+// Usage example
+// namespace prc = prism::connect;
+// void configCallback(prc::Client& client)
+// {
+//     client.setLogFlags(prc::Client::LOG_INPUT | prc::Client::LOG_INPUT_JSON);
+//     client.setConnectionTimeoutMs(5000);
+//     client.setLowSpeed(5);
+// }
+//
+// prc::ArtifactUploader::Configuration uploaderConfig("<apiRoot>", "<apiToken>",
+//      "camOne", 32);
+// prc::ArtifactUploader uploader;
+// prc::Status status = uploader.init(uploaderConfig, configCallback);
+// if (status.isError())
+//     abort();
+// uploader.uploadBackground(<timestamp>, PayloadAu::makeByMovingData(<data>, <mimeType>));
+
+//class PayloadAu;
+//typedef boost::shared_ptr<PayloadAu> PayloadAuPtr;
+
+// TODO: other options for class name: SmartPayload, PayloadV2, PayloadTwo
+// This class is carefully designed to take ownership and clean data
+// after use.
+// The moment, when data is copied or stolen is well-defined: in a call
+// to make* static method.
+// The moment, when owned data or file is deleted is uknknown: whenever
+// PayloadAu instance is destroyed.
+// Copying and assignment are explicitly prohibited to prevent unintended
+// losing of ownership or data copying.
+// This class is intentionally made distinct from Payload struct to keep
+// current interface using Payload unchanged.
+class PayloadAu : private boost::noncopyable
+{
+public:
+    static PayloadAuPtr makeByMovingData(move_ref<ByteBuffer> data, const std::string& mimeType);
+    static PayloadAuPtr makeByCopyingData(const void* data, size_t dataSize, const std::string& mimeType);
+
+    // This method is intentionally not implemented, as it isn't good for
+    // async interface
+    // PayloadAuPtr makeByReferencingData();
+
+    // This method is intentionally not implemented, as it isn't good for
+    // async interface
+    // PayloadAuPtr makeByReferencingFile();
+
+    // file will be deleted on PayloadAu destruction
+    static PayloadAuPtr makeByReferencingFileAutodelete(const std::string& filePath);
+
+    bool isFile() const;
+    std::string getFilePath() const;
+    std::string getMimeType() const;
+    const void* getData() const;
+    size_t getDataSize() const;
+
+private:
+    PayloadAu();
+
+    class Impl;
+    unique_ptr<Impl>::t pImpl_;
+};
+
+class ArtifactUploader
+{
+public:
+    typedef void (*ClientConfigCallback)(Client& client);
+
+    struct Configuration
+    {
+        Configuration(const std::string& apiRoot,
+                      const std::string& apiToken,
+                      const std::string& cameraName,
+                      int maxQueueSizeMB,
+                      const std::string& queueType = "simple",
```

Comment:
The const reference to a temporary is OK, but I would prefer to have overloaded constructor or pass type by value instead. Just to be safe.